### ISSUE TITLE
Migration to Swift

### ISSUE THREAD

**Issue by YuriyVelichkoOpenX** [prebid/prebid-mobile-ios#396]
## Motivation

The public classes of the Rendering Module were migrated from Objective-C to Swift. All others are on Objective-C still. 

Swift Package Manager does not support mixed frameworks. Therefore we can't provide Prebbid Mobile with rendering through SPM. 

We could split Objective-C and Swift parts to introduce different modules, but it controversies our modularization strategy (#395) in part dedicated to following the principles of layered architecture. So it is better to concentrate efforts on the complete migration.

## Given

In the following doc, you can find the list of classes for migration

https://docs.google.com/spreadsheets/d/1ZZoRkXsIa95RZELCoEsP4IHJ5TDyRsNAlfKSQU2YohI/edit?usp=sharing

## Objective

Feel free to take any Objective-C class, mark the row in the doc with yellow color, and migrate it to Swift. 

Try to follow the priority column. It will help to take the less coupled classes first. 

Mark migrated classes green after migration.

**Comment by YuriyVelichkoPI** [comment]
The implementation of this epic will fix SPM integration

**Comment by YuriyVelichkoPI** [comment]
Posting a phased migration plan drafted with Claude based on a fresh codebase audit. Reopening this issue — the migration is not complete: `PrebidMobile/Objc/` still contains ~119 `.m` and 152 `.h` files (vs 271 already-Swift files).

---

# Migration plan: PrebidMobile iOS SDK → Swift

## 1. Goal & scope

Migrate the entire ObjC layer in `PrebidMobile/Objc/` (~119 `.m` + 152 `.h` files) to Swift in independently mergeable steps. The Swift API layer in `PrebidMobile/Swift/` (271 files) is unchanged in shape — Swift twins land under the same path on the Swift side.

In scope: everything under `PrebidMobile/Objc/` including ORTB models, networking, AdView/MRAID/VAST rendering, OMSDK wrapper, utilities.
Out of scope: `EventHandlers/` adapters (already mostly Swift), tests don't need Swift rewrites unless a class change forces it.

## 2. Constraints

- **Min iOS deployment target stays at 13.0.** Any step that would require a bump must stop and ask first.
- **Drop ObjC consumer support.** The public umbrella `PrebidMobile.h` is empty — all `PBM*` ObjC types are public only to cross the SPM split-target seam (`__PrebidMobileInternal` ↔ `PrebidMobile`). No external app imports them. After migration, types collapse back to `internal`.
- **No public-API behavior changes.** The Swift surface (`Prebid`, `Targeting`, ad units, `PrebidAdUnit`, etc.) keeps exact same semantics. Bid request bytes on the wire must be byte-for-byte equivalent for the Phase-1/3 steps.
- **Keep CocoaPods + SPM + Carthage build paths green at every step.** Each step ends with a green `./scripts/buildPrebidMobile.sh` + `./scripts/testPrebidMobile.sh --latest --quick`.

## 3. Strategy

**Per-class file-level twin.** For each `Foo.m` + `Foo.h` (+ optional `Foo+Protected.h`):

1. Create `Foo.swift` at the mirrored path under `PrebidMobile/Swift/...`.
2. Adopt the existing Swift base — `PBMJsonCodable` (`= PBMJsonDecodable & PBMJsonEncodable`) at `PrebidMobile/Swift/PrebidMobileRendering/Prebid/PBMCore/PBMJsonCoding.swift`. The response-side ORTB Swift code (`Prebid/PBMCore/ORTB/ORTBBid.swift` etc.) is the canonical pattern.
3. Port property declarations 1:1, preserving exact JSON keys.
4. Port `initWithJsonDictionary:` → `init?(jsonDictionary:)`. Port `toJsonDictionary` → `var jsonDictionary: [String: Any]`. Port `fromJsonString:`/`toJsonStringWithError:` are inherited from the protocol's default implementations — do not reimplement per class.
5. Replace `PBMJsonDictionary` typedef with `[String: Any]` Swift-native at every site touched.
6. Replace `<NSCopying>` with an explicit `func copy() -> Self` only when a real callsite exists. Audit before copying the conformance.
7. Update Swift call sites that referenced the ObjC type via the auto-generated `-Swift.h` to use the new Swift type directly. Update remaining ObjC call sites to import the Swift module's `*-Swift.h`.
8. Delete `Foo.m`, `Foo.h`, and any matching entry in `PrebidMobile/Objc/PrivateHeaders/`.
9. Verify: quick tests + full tests for the affected scheme (`PrebidMobileTests` for core, `PrebidMobileGAMEventHandlersTests` etc. for adapters when networking touched).

**Cross-step rule.** Never delete an ObjC type whose consumers haven't been re-pointed in the same PR. If the dependency graph forces a bigger PR, group those classes into one step rather than splitting.

**Functional-gap audit at Phase 1 entry.** Before migrating any PBMORTB type, diff `PBMORTBAbstract.h` + `+Protected.h` against `PBMJsonCoding.swift`. Known gaps to evaluate: (a) `<NSCopying>` deep-copy semantics — replace per class only if used; (b) thread safety — `CustomModelObjects` already uses `pthread_rwlock`. If a gap exists, extend `PBMJsonCoding.swift` (or an `ORTBJsonCodable` refinement protocol) once, then all subsequent ORTB twins use it. **Don't fork the base.**

## 4. Validation per step

- **Compile parity:** `xcodebuild -workspace PrebidMobile.xcworkspace -scheme Lib-PrebidMobile -sdk iphonesimulator build` and the framework build script.
- **Test parity:** `./scripts/testPrebidMobile.sh --latest --quick` minimum; `--latest` (full) on the final PR of each phase.
- **JSON round-trip parity (Phase 1 & 3 only):** add an XCTest that loads a fixture `BidRequest.json` and a `BidResponse.json`, decodes through the ObjC type (snapshot from `master`) and the new Swift type, and asserts `jsonDictionary` equality. Delete the ObjC fixture leg once migration is done.
- **Bid-on-the-wire parity (Phase 3 final):** record one bid request body on `master`, replay the same `Prebid.fetchDemand(...)` call on the migration branch, diff JSON. Must match.

## 5. Phasing & ordering

Phases are ordered by risk (low → high). Steps within a phase can be parallelized when their "Depends on" lists don't intersect. Each step is **one PR**.

---

### Phase 0 — Setup (no migration)

**S0.1 — Functional-gap audit & playbook codification.** *Single, must merge first.*
- Read `PBMORTBAbstract.h`, `+Protected.h`, `PBMORTB_NotImplemented.h`, `PBMConstants.h`, `PBMJsonDictionary` typedef.
- Compare to `PBMJsonCoding.swift`. Document any extension needed (e.g. NSCopying replacement, NSError-out-param parity).
- Produce `docs/migration/playbook.md` codifying §3.
- **Deliverable:** doc + (if any) one-line additions to `PBMJsonCoding.swift`. No `.m` deletions.
- **Risk:** low. **Size:** S.

**S0.2 — JSON round-trip parity harness.** *Optional but strongly recommended.*
- Add `PrebidMobileTests/Helpers/ORTBParityHelper.swift` with `assertJSONParity(objc:swift:)`.
- Add fixture JSONs to `PrebidMobileTests/Resources/Migration/`.
- **Deliverable:** harness + 1-2 baseline fixtures.
- **Risk:** low. **Size:** S.

---

### Phase 1 — ORTB bid-request models (`PrebidMobile/Objc/PrebidMobileRendering/ORTB/`, 25 files)

The bid-RESPONSE side is already Swift (`Prebid/PBMCore/ORTB/`). The bid-REQUEST side is the goal here. Risk is low — these are data carriers.

**S1.1 — ORTB leaf models.**
- Files in: `PBMORTBFormat`, `PBMORTBPublisher`, `PBMORTBGeo`, `PBMORTBDeal`, `PBMORTBSourceExtOMID`, `PBMORTBImpExtSkadn`, `PBMORTBDeviceExtAtts`.
- New Swift twins under `PrebidMobile/Swift/PrebidMobileRendering/ORTB/Request/`.
- **Depends on:** S0.1.
- **Risk:** low. **Size:** M.

**S1.2 — Composite request blocks.**
- Files in: `PBMORTBBanner`, `PBMORTBVideo`, `PBMORTBPmp`, `PBMORTBImpExtPrebid`, `PBMORTBImp`.
- **Depends on:** S1.1 (PBMORTBImp uses Banner/Video/Pmp; Pmp uses Deal).
- **Risk:** low. **Size:** M.

**S1.3 — Top-level objects.**
- Files in: `PBMORTBApp`, `PBMORTBAppExt`, `PBMORTBAppExtPrebid`, `PBMORTBDevice`, `PBMORTBDeviceExtPrebid`, `PBMORTBDeviceExtPrebidInterstitial`, `PBMORTBUser`, `PBMORTBRegs`, `PBMORTBSource`, `PBMORTBRendererConfig`.
- **Depends on:** S1.1, S1.2 (Device/App use Geo/Format/Publisher).
- **Risk:** low. **Size:** L — split into S1.3a (App+Device family) and S1.3b (User+Regs+Source+RendererConfig) if review burden is heavy.

**S1.4 — Root container + parameter builder.**
- Files in: `PBMORTBBidRequest`, `PBMORTBBidRequestExtPrebid`, `PBMORTBAbstract` (delete after no consumers remain), `PBMORTBParameterBuilder` (the ORTB-layer one).
- **Depends on:** S1.1–S1.3.
- **Risk:** medium — touches the request-assembly path; consumers in Phase 3 will need updating.
- **Acceptance:** §4 bid-on-the-wire parity test passes.

---

### Phase 2 — Utilities & extensions (parallel-safe with Phase 1)

`PrebidMobile/Objc/PrebidMobileRendering/Utilities/`, `ExtensionsAndWrappers/`, `Objc/Utils/`. ~24 files total. No deps on ORTB.

**S2.1 — Standalone utilities.** `PBMConstants`, `PBMFunctions`, `PBMMacros`, `PBMLog`, etc. (`Utilities/`, 12 files). **Risk:** low. **Size:** M. Split into 2-3 PRs by file group if needed.

**S2.2 — Foundation category extensions.** `NSDictionary+PBMExtensions`, `NSMutableDictionary+PBMExtensions`, `NSString+PBMExtensions`, etc. (`ExtensionsAndWrappers/`, 8 files; `Exposure/`, 3 files). **Risk:** low. **Size:** M.

**S2.3 — NSTimer wrapper.** 2 files. **Risk:** low. **Size:** S.

---

### Phase 3 — Networking parameter builders + transport (`Networking/`, ~12 files)

Consumes Phase 1 (PBMORTB types) and Phase 2 (utilities).

**S3.1 — Parameter builders (group, tightly coupled).**
- Files in: `PBMBasicParameterBuilder`, `PBMNetworkParameterBuilder`, `PBMUserConsentParameterBuilder`, `PBMGeoLocationParameterBuilder`, `PBMAppInfoParameterBuilder`, `PBMDeviceInfoParameterBuilder`, `PBMSKAdNetworksParameterBuilder`, `PBMORTBParameterBuilder` (the Networking-layer one, distinct from ORTB-layer).
- **Depends on:** S1.4.
- **Risk:** medium.
- **Acceptance:** §4 bid-on-the-wire parity test still green.

**S3.2 — Builder orchestrator.** `PBMParameterBuilderService`. **Depends on:** S3.1. **Risk:** low. **Size:** S.

**S3.3 — URL building.** `Networking/URLBuilding/` (1 file). **Risk:** low. **Size:** S.

**S3.4 — Impression tracking.** `Networking/Tracking/` (1 file). **Risk:** low. **Size:** S.

---

### Phase 4 — PBMCore (`PrebidMobileRendering/Prebid/PBMCore/`, ~13 files)

The orchestration layer. Touches request/response, native events, transactions.

**S4.1 — Bid response transformer.** `PBMBidResponseTransformer`. Bridges ObjC ORTB response → existing Swift `BidResponse`. Once Swift, may be merged with `BidResponse.swift`. **Depends on:** S1.4. **Risk:** medium. **Size:** M.

**S4.2 — Bid requester.** `PBMBidRequester`. **Depends on:** S3.x, S4.1. **Risk:** medium-high (most-used class in the SDK). **Size:** M.

**S4.3 — Prebid parameter builder.** `PBMPrebidParameterBuilder`. **Depends on:** S3.x. **Risk:** medium. **Size:** M.

**S4.4 — Native click tracking.** `PBMCore/NativeEventsTracking/ClickTracking/` (4 files). **Risk:** medium. **Size:** M.

**S4.5 — Transaction factory.** `PBMCore/TransactionFactory/` (3 files). **Depends on:** S4.2. **Risk:** medium. **Size:** M.

---

### Phase 5 — AdView rendering (non-WebView path)

`AdTypes/AdView/`, `CreativeFactory/`. ~15 files.

**S5.1 — Ad load manager.** `AdView/AdLoadManager/` (2 files). **Risk:** medium. **Size:** S.

**S5.2 — AdView core (group).** Files under `AdView/` excluding modals — coupled enough to migrate together. ~8 files. **Risk:** medium. **Size:** L.

**S5.3 — Modals.** `AdView/Modals/` (2 files). **Depends on:** S5.2. **Risk:** medium. **Size:** S.

**S5.4 — Creative factory.** `CreativeFactory/` (2 files). **Depends on:** S5.2. **Risk:** medium. **Size:** S.

---

### Phase 6 — VAST / Video (highest data-parser risk)

`AdTypes/Video/Vast/`, 15 files. Complex parser + player integration.

**S6.1 — VAST primitives (group).** Leaf parser nodes: `Ad`, `Creative`, `Linear`, `MediaFile`, tracking event types, etc. **Risk:** high. **Size:** L.

**S6.2 — VAST root + ad selector.** `PBMVastParser`, `PBMVastResponse`. **Depends on:** S6.1. **Risk:** high. **Size:** M.

**S6.3 — Video player integration.** `PBMVideoCreative` + glue with AVKit. **Depends on:** S5.x, S6.2. **Risk:** high. **Size:** M.

---

### Phase 7 — HTML / MRAID (highest overall risk)

`AdTypes/AdView/` MRAID & HTML rendering pieces, plus the `PBMAbstractCreative` base. WKWebView + JS bridge.

**S7.1 — MRAID JS bridge.** Message channel + dispatcher. **Risk:** very high. **Size:** L.

**S7.2 — HTML creative rendering.** `PBMHTMLCreative` and friends. **Depends on:** S7.1. **Risk:** very high. **Size:** L.

**S7.3 — Abstract creative base.** `PBMAbstractCreative`. **Depends on:** S5.4, S7.2, S6.3. **Risk:** high (root of the rendering hierarchy). **Size:** M.

---

### Phase 8 — 3rd-party wrapper (`3dPartyWrappers/OpenMeasurement/`, 4 files)

OMSDK is shipped as a bundled `.xcframework` (`Frameworks/OMSDK_Prebidorg.xcframework`). The wrapper itself is thin.

**S8.1 — OMSDK wrapper.** **Risk:** medium (compliance — OM viewability measurement must keep working; verify with OM SDK validator). **Size:** M. **Defer until Phase 7 stable.**

---

### Phase 9 — Cleanup & SPM seam collapse

**S9.1 — Header & typedef sweep.** Delete remaining files in `Objc/PrivateHeaders/`. Remove `PBMJsonDictionary` typedef. Remove `<NSCopying>` declarations that have no callsites. **Risk:** low. **Size:** M.

**S9.2 — Visibility tightening.** Demote `public` → `internal` for types that were public only for the SPM ObjC↔Swift seam. Run `swift build` for the SPM target to confirm. **Risk:** low. **Size:** M.

**S9.3 — SPM target collapse (consider).** With no ObjC remaining, `__PrebidMobileInternal` becomes empty. Either keep the structure with a placeholder or merge into `PrebidMobile`. Check that this doesn't break consumers using `import PrebidMobile`. **Risk:** medium. **Size:** S.

**S9.4 — CLAUDE.md / README update.** Update architecture section to remove the Swift/ObjC split description. **Risk:** none. **Size:** S.

## 6. Step ID quick reference

```
Phase 0:  S0.1, S0.2
Phase 1:  S1.1, S1.2, S1.3 (a/b), S1.4
Phase 2:  S2.1, S2.2, S2.3                  (parallel with Phase 1)
Phase 3:  S3.1, S3.2, S3.3, S3.4
Phase 4:  S4.1, S4.2, S4.3, S4.4, S4.5
Phase 5:  S5.1, S5.2, S5.3, S5.4
Phase 6:  S6.1, S6.2, S6.3
Phase 7:  S7.1, S7.2, S7.3
Phase 8:  S8.1
Phase 9:  S9.1, S9.2, S9.3, S9.4
```

~32 PRs total. Phases 1+2 can run in parallel. Phase 3 cannot start until S1.4 lands. Phases 6 / 7 are gated by Phase 5.

## 7. Open assumptions to revisit before each phase

- ObjC `<NSCopying>` is rarely actually invoked — verify per class with grep before dropping.
- `PBMORTBAbstract.fromJsonString:` callers in tests use the throwing Swift form post-migration. Audit during S1.4.
- Whether bid-request and bid-response Swift hierarchies should ultimately share folder structure (`Swift/.../ORTB/{Request,Response}/`). Decide at S1.1; can refactor later.
- Whether the empty `__PrebidMobileInternal` SPM target can be collapsed without breaking downstream consumers. Decide at S9.3.

## 8. Tracking

Each step (S0.1, S0.2, …) becomes its own GitHub PR titled `[swift-migration] <step ID> <one-line scope>`. The first PR links back to this issue; subsequent PRs reference the previous one.

---

_Plan drafted with Claude (Opus 4.7, 1M context) based on a fresh codebase audit. File counts and class names reflect repo state on branch `1259-Fix-converting-of-decimal-numbers-in-setGlobalORTBConfig` as of 2026-04-28._
